### ISSUE TITLE
Optimize `call_memset` etc. to handle small aligned fixed-size cases

### ISSUE THREAD

**Issue by sunfishcode** [CraneStation/cranelift#542]
`FunctionBuilder`'s new `call_memcpy` currently always generates a call to "memcpy", which is correct, however it's very common in Rust code to need to make copies of small objects with known sizes and alignments, which can be more quickly done with a handful of loads and stores, rather than making a call.

We should introduce additional `FunctionBuilder` functions, with names along the lines of `emit_small_memcpy`, which also take source and destination pointer `Value`s, but instead of a length `Value`, have a `u64` specifying the size in bytes. It should also take two additional `u8` parameters, specifying the alignments in bytes of the source and destination pointers (and required to be powers of two).

For the first step, we can just ignore the alignments, put the size into a register using `iconst`, and then use `call_memcpy`. That will be the conservative fallback.

The next step is the main load/store optimization.
 - First, determine the access type for the loads and stores.
    - The alignments for the source and destination will often be the same, particularly in the case of struct copying, so for now we can focus on that case. So we start by computing the minimum of the source and destination values.
    - Then compute the greatest power of two that evenly divides the size and that's at most that minimum alignment value. That will be the size of the loads and stores we emit.
    - Next, divide the length by that size to determine how many loads and stores we'll need. If it's "too many", then we should use the conservative fallback rather than emitting a huge number of instructions. For now, let's say "too many" is 4 loads and 4 stores (this is a wild guess, but we canl tune it later).
    - Finally, emit loads and stores using, picking an integer type based on the access size we computed. We can use the constant offset fields of `load` and `store` so that we can always just use the base source and destination `Value`s without having to do address arithmetic. And we should interleave the loads and stores (load, store, load store, rather than load, load, store store) since that allows the register allocator to reuse a register rather than requiring all the values get their own register.

That's `memcpy`. `memmove` is similar, but the source and destination can overlap. So for memmove, we *don't* want to interleave the loads and stores. If we do all the loads first, then all the stores second, we correctly handle the overlap case. We'll need more registers that way, but that's the cost of using memmove instead of memcpy :-).

Then, `memset` is similar, except there, we should replace the byte `Value` operand with a `u8` value. We can then materialize it into an `iconst` with the value (c | (c << 8) | (c << 16) ...) out to the destination alignment width, and then do stores with that.

**Comment by lachlansneff** [comment]
Perhaps, we should just optimize this in the constant folding pass in `cranelift-preopt`.

**Comment by sunfishcode** [comment]
Yes, we may eventually want to add a `memcpy` opcode to the IR, and optimize it like that. It should be easy to transition to doing that, because we can just make the cranelift-frontend APIs emit that instruction rather than immediately lowering it, and move the lowering code into cranelift-codegen.

My thinking for lowering it in cranelift-frontend right now is mainly to keep codegen a little simpler for now, while we're still working on some larger design questions. However, it's not a hard rule.

The Rust compiler, for its part, usually knows when it's doing fixed-sized copies, so it shouldn't be a trouble for it to emit these directly.

**Comment by lachlansneff** [comment]
Okay, that all sounds fine. It should definitely be possible to emit different different memcpy implementations depending on the target. For general use, just call the libc implementation, but for copies that are known to be quite large, it could emit `rep movsb`, etc.
